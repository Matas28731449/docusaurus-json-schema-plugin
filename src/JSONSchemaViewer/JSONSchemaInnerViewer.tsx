import React from "react"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  schema: JSONSchema7
}

function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <></>
}

export default JSONSchemaInnerViewer
