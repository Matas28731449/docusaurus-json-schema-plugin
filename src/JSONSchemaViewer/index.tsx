import React, { useEffect, useState } from "react"
import Layout from "@theme/Layout"
import Translate from "@docusaurus/Translate"
import AllOfMerger from "json-schema-merge-allof"
import { Resolver } from "@stoplight/json-ref-resolver"

import type { JSONSchema7 } from "json-schema"

import JSONSchemaViewer from "./JSONSchemaInnerViewer"

type ViewerProperties = {
  loadSchema: () => Promise<JSONSchema7>
}

// merged representation of allOf array of schemas
async function mergeAllOf(
  schema: JSONSchema7
): Promise<Omit<JSONSchema7, "allOf">> {
  const mergedSchema = AllOfMerger(schema, {
    // Technically valid, but only interested on fields with name, at least for now ...
    ignoreAdditionalProperties: true,
  })

  return Promise.resolve(mergedSchema as Omit<JSONSchema7, "allOf">)
}

function JSONSchemaEditor(props: ViewerProperties): JSX.Element {
  const [schema, setSchema] = useState(
    undefined as undefined | Omit<JSONSchema7, "allOf">
  )
  const [fetchError, setFetchError] = useState(undefined as undefined | Error)

  useEffect(() => {
    props
      .loadSchema()
      .then((userSchema) => new Resolver().resolve(userSchema))
      .then((resolvedSchema) => mergeAllOf(resolvedSchema.result))
      .then((schemaWithoutAllOf) => setSchema(schemaWithoutAllOf))
      .catch((err) => setFetchError(err))
  }, [])

  return (
    <Layout>
      {fetchError !== undefined && (
        <div>
          <p>
            <Translate
              values={{
                errorMessage: fetchError.message,
                id: "docusaurus-json-schema-viewer-plugin.error",
              }}
            >
              {"This component crashed because of error: {errorMessage}."}
            </Translate>
          </p>
        </div>
      )}
      {schema === undefined && (
        <div>
          <p>
            <Translate values={{ id: "docusaurus-json-schema-viewer-plugin.loading" }}>
              {"Loading schema ..."}
            </Translate>
          </p>
        </div>
      )}
      {schema !== undefined && <JSONSchemaViewer schema={schema} />}
    </Layout>
  )
}

export default JSONSchemaEditor
