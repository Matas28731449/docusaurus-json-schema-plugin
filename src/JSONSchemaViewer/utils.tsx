import React from "react";

import type { JSONSchema7 } from "json-schema";

// To keep the same type everywhere here
type cleanSchema = Omit<JSONSchema7, "allOf">;

