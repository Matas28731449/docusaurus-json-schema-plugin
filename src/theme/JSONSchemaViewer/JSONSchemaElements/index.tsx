// Component to deal with oneOf / anyOf / allOf / not
export { SchemaComposition } from "./schemaComposition/index"

// Component to deal with if-then-else , dependentRequired , dependentSchemas , dependencies
export { SchemaConditional } from "./SchemaConditional/index"

// Component to deal with Object
export { CreateObject } from "./object/index"

// Component to deal with Array
export { CreateArray } from "./array/index"

// Component to deal with String
export { default as CreateString } from "./CreateString"

// Component to deal with Boolean
export { default as CreateBoolean } from "./CreateBoolean"

// Component to deal with Number
export { default as CreateNumber } from "./CreateNumber"

// Component to deal with Integer
export { default as CreateInteger } from "./CreateInteger"

// Component to deal with schema defined as "true"
export { default as CreateAlwaysValid } from "./CreateAlwaysValid"

// Component to deal with schema defined as "false"
export { default as CreateAlwaysInvalid } from "./CreateAlwaysInvalid"
