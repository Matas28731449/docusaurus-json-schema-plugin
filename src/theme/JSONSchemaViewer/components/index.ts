// To avoid circular dependancies issue, let's have a single place for exports and imports

// Collapsible componenent
export { default as Collapsible } from "./Collapsible"

// CreateDetailsNode component
export { default as CreateDetailsNode } from "./CreateDetailsNode"

// CreateEdges component
export { default as CreateEdges } from "./CreateEdges"

// CreateProperties component
export { default as CreateProperties } from "./CreateProperties"

// RenderAnyOneOf component
export { default as RenderAnyOneOf } from "./RenderAnyOneOf"

// https://stackoverflow.com/a/69328045/6149867
export type WithRequired<T, K extends keyof T> = T & { [P in K]-?: T[P] }
