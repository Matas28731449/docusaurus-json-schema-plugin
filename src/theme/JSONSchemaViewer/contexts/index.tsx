export * from "@theme/JSONSchemaViewer/contexts/jsvOptions"
export * from "@theme/JSONSchemaViewer/contexts/schemaHierarchy"
