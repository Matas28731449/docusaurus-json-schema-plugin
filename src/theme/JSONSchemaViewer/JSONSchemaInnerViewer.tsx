import React from "react"
import { CreateNodes } from "./components/index"

import type { JSONSchema7 } from "json-schema"

type InnerViewerProperties = {
  // Thanks to json-schema-merge-allof , we don't have allOf in the whole user schema
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: Omit<JSONSchema7, "allOf">
}

// External
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  return <ul style={{ marginLeft: "1rem" }}>{CreateNodes(props.schema)}</ul>
}

export default JSONSchemaInnerViewer
