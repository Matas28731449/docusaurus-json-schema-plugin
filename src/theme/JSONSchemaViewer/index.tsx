import React, { useState, useEffect } from "react"
import { Resolver } from "@stoplight/json-ref-resolver"

import { CreateNodes, Collapsible } from "./components/index"
import { JSVOptionsContextProvider } from "./contexts/index"

import type { JSONSchema } from "./types"
import type { JSVOptions } from "./contexts/index"
import type { IResolveOpts } from "@stoplight/json-ref-resolver/types"

export type Props = {
  // The schema to use
  schema: unknown
  // To customize the ref resolving
  resolverOptions?: IResolveOpts
  // To customize the viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
  // Other properties (who knows ?)
  [x: string]: any
}

type InnerViewerProperties = {
  // Thanks to @stoplight/json-ref-resolver, $ref are either :
  // 1. resolved
  // 2. unresolved (as circular stuff are not on the roadmap)
  schema: JSONSchema
  // Options for viewer
  viewerOptions?: Omit<JSVOptions, "fullSchema">
}

// Internal
function JSONSchemaInnerViewer(props: InnerViewerProperties): JSX.Element {
  const { schema, viewerOptions } = props
  // Title of the schema, for user friendliness
  const title =
    typeof schema !== "boolean" && schema?.title !== undefined
      ? schema.title
      : "Schema"

  // state for provider
  const startingState: JSVOptions = {
    fullSchema: schema,
    // spread provided attributes
    ...viewerOptions,
  }

  return (
    <JSVOptionsContextProvider value={startingState}>
      <Collapsible
        summary={<strong>{title}</strong>}
        detailsProps={{
          open: true,
        }}
      >
        <CreateNodes schema={schema} />
      </Collapsible>
    </JSVOptionsContextProvider>
  )
}

// Entry point
export default function JSONSchemaViewer(props: Props): JSX.Element {
  const { schema: originalSchema, resolverOptions, viewerOptions } = props

  const [error, setError] = useState(undefined as undefined | Error)
  const [resolvedSchema, setResolvedSchema] = useState(
    undefined as undefined | JSONSchema
  )

  useEffect(() => {
    // Time to do the job
    new Resolver()
      .resolve(originalSchema, resolverOptions)
      .then((result) => {
        setResolvedSchema(result.result)
      })
      .catch((err) => {
        setError(err)
      })
  }, [])

  if (error !== undefined) {
    return <div>Something bad happens : {error.message}</div>
  } else if (resolvedSchema === undefined) {
    return <div>Loading ....</div>
  } else {
    return (
      <JSONSchemaInnerViewer
        schema={resolvedSchema}
        viewerOptions={viewerOptions}
      />
    )
  }
}
