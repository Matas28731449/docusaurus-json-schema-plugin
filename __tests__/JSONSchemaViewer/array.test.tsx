import React from "react"

// For typings autocomplete whatever your IDE
import { expect, test, describe } from "@jest/globals"

import { create, act } from "react-test-renderer"

import JSONSchemaViewer from "../../src/theme/JSONSchemaViewer/index"

// Type to prevent creating invalid mocks
import type { JSONSchema } from "../../src/theme/JSONSchemaViewer/types"

// Type for react-test-renderer
import type { ReactTestRenderer } from "react-test-renderer"

const testcases: JSONSchema[] = [
  {
    type: "array",
  },
  {
    type: "array",
    contains: {
      type: "number",
    },
  },
  {
    type: "array",
    items: {
      type: "number",
    },
  },
  {
    type: "array",
    minItems: 2,
    maxItems: 3,
  },
  {
    type: "array",
    contains: {
      type: "integer",
    },
    minContains: 2,
    maxContains: 3,
  },
  {
    type: "array",
    description:
      "Represent a street address such as ['1600','Pennsylvania','Avenue','NW']",
    items: false,
    prefixItems: [
      {
        type: "number",
        description: "The address number",
      },
      {
        type: "string",
        description: "The name of the street",
      },
      {
        enum: ["Street", "Avenue", "Boulevard"],
        description: "The type of street",
      },
      {
        enum: ["NW", "NE", "SW", "SE"],
        description: "The city quadrant of the address",
      },
    ],
  },
  {
    type: "array",
    uniqueItems: true,
  },
  {
    minItems: 1,
  },
  {
    maxItems: 5,
  },
  {
    maxContains: 3,
  },
  {
    minContains: 2,
  },
  {
    type: "array",
    items: [{ type: "integer" }, { type: "string" }],
    minItems: 1,
  },
  {
    type: "array",
    prefixItems: [{ type: "number" }, { type: "string" }],
    minItems: 3,
  },
  {
    type: "array",
    items: {
      type: "string",
    },
    minItems: 1,
  },
  {
    unevaluatedItems: false
  }
]

describe("JSONSchemaViewer - Array type", () => {
  test.each(testcases)("test %#", async (fakeSchema) => {
    // render the component
    let root: ReactTestRenderer | undefined

    await act(async () => {
      root = create(<JSONSchemaViewer schema={fakeSchema} />)
    })

    // make assertions on root
    expect(root?.toJSON()).toMatchSnapshot()
  })
})
