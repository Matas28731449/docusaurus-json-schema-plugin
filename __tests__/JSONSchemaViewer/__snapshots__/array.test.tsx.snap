// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JSONSchemaViewer - Array type test 0 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - Array type test 1 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            contains
          </code>
          Â 
          <span>
            number
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          number
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - Array type test 2 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items
          </code>
          Â 
          <span>
            number
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          number
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  />
</details>
`;

exports[`JSONSchemaViewer - Array type test 3 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      <strong>
        Length :
      </strong>
      Â 
      <code>
        &gt;= 2
      </code>
      Â 
      AND
      Â 
      <code>
        &lt;= 3
      </code>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - Array type test 4 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            contains
          </code>
          Â 
          <span>
            integer
          </span>
          Â 
          <strong>
            required
          </strong>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          integer
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      <strong>
        Must contain : 
      </strong>
      Â 
      <code>
        at least 2 valid item(s)
      </code>
      Â 
      AND
      Â 
      <code>
        at most 3 valid item(s)
      </code>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - Array type test 5 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <ul />
  <ul>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[0]
          </code>
          Â 
          <span>
            number
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          number
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          The address number
        </div>
      </details>
    </li>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[1]
          </code>
          Â 
          <span>
            string
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          string
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        />
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          The name of the street
        </div>
      </details>
    </li>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[2]
          </code>
          Â 
          <span>
            unknown
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          unknown
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          <div>
            <strong>
              Possible values :
            </strong>
            Â 
            <code>
              ["Street","Avenue","Boulevard"]
            </code>
          </div>
        </div>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          The type of street
        </div>
      </details>
    </li>
    <li>
      <details
        open={false}
      >
        <summary>
          <code>
            items[3]
          </code>
          Â 
          <span>
            unknown
          </span>
        </summary>
        <strong>
          type
        </strong>
        Â :Â 
        <span
          style={
            {
              "opacity": "0.6",
            }
          }
        >
          unknown
        </span>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          <div>
            <strong>
              Possible values :
            </strong>
            Â 
            <code>
              ["NW","NE","SW","SE"]
            </code>
          </div>
        </div>
        <div
          style={
            {
              "marginTop": "var(--ifm-table-cell-padding)",
            }
          }
        >
          The city quadrant of the address
        </div>
      </details>
    </li>
  </ul>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      ðŸš¨Â 
      No extra item(s) are authorized in this array
    </div>
  </div>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    Represent a street address such as ['1600','Pennsylvania','Avenue','NW']
  </div>
</details>
`;

exports[`JSONSchemaViewer - Array type test 6 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      <strong>
        Unique items :
      </strong>
      Â 
      <code>
        yes
      </code>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - Array type test 7 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      <strong>
        Length :
      </strong>
      Â 
      <code>
        &gt;= 1
      </code>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - Array type test 8 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      <strong>
        Length :
      </strong>
      Â 
      <code>
        &lt;= 5
      </code>
    </div>
  </div>
</details>
`;

exports[`JSONSchemaViewer - Array type test 9 1`] = `
<details
  open={true}
>
  <summary>
    <strong>
      Schema
    </strong>
  </summary>
  <strong>
    type
  </strong>
  Â :Â 
  <span
    style={
      {
        "opacity": "0.6",
      }
    }
  >
    array
  </span>
  <div
    style={
      {
        "marginTop": "var(--ifm-table-cell-padding)",
      }
    }
  >
    <div>
      <strong>
        Must contain : 
      </strong>
      Â 
      <code>
        at most 3 valid item(s)
      </code>
    </div>
  </div>
</details>
`;
